tests/framework/utils.py

```diff
@@ -91,25 +92,73 @@ def get_cpu_percent(pid: int) -> Dict[str, Dict[str, float]]:
 class UffdHandler:
     """Describe the UFFD page fault handler process."""
 
-    def __init__(self, name, args):
+    def __init__(self, name, socket_path, mem_file, chroot, log_file_name):
         """Instantiate the handler process with arguments."""
         self._proc = None
-        self._args = [f"/{name}"]
-        self._args.extend(args)
+        self._handler_name = name
+        self._socket_path = socket_path
+        self._mem_file = mem_file
+        self._chroot = chroot
+        self._log_file = log_file_name
 
-    def spawn(self):
+    def spawn(self, uid, gid):
         """Spawn handler process using arguments provided."""
-        self._proc = subprocess.Popen(
-            self._args, stdout=subprocess.PIPE, stderr=subprocess.PIPE
-        )
 
+        # Need to keep these around so we can exit the chroot
+        real_root = os.open("/", os.O_RDONLY)
+        working_dir = os.getcwd()
+
+        # Jump in the chroot
+        os.chroot(self._chroot)
+        os.chdir("/")
+        st = os.stat(self._handler_name)
+        os.chmod(self._handler_name, st.st_mode | stat.S_IEXEC)
+
+        chroot_log_file = Path("/") / self._log_file
+        with open(chroot_log_file, "w", encoding="utf-8") as logfile:
+            args = [f"/{self._handler_name}", self._socket_path, self._mem_file]
+            self._proc = subprocess.Popen(
+                args, stdout=logfile, stderr=subprocess.STDOUT
+            )
+
+        time.sleep(1)
+        if not self.is_running():
+            print(chroot_log_file.read_text(encoding="utf-8"))
+            assert False, "Could not start PF handler!"
+
+        # The page fault handler will create the socket path with root rights.
+        # Change rights to the jailer's.
+        os.chown(self._socket_path, uid, gid)
+
+        os.fchdir(real_root)
+        os.chroot(".")
+        os.chdir(working_dir)
```

Comment:
Could do a context manager for the chroot part:

```python
@contextmanager
def chroot(*args, **kwds):
    try:
        # Need to keep these around so we can exit the chroot
        real_root = os.open("/", os.O_RDONLY)
        working_dir = os.getcwd()

        # Jump in the chroot
        os.chroot(self._chroot)
        os.chdir("/")
        yield
    finally:
        os.fchdir(real_root)
        os.chroot(".")
        os.chdir(working_dir)
```

Then use it like:

```
with chroot("/chroot"):
    pass

```

Reply:
That's awesome. I wanted to do something like that, I failed and then gave up, given that we only seem to use it in this particular test. I 'll give it a go.